### ISSUE TITLE
Bad WER for new dataset

### ISSUE THREAD

**Issue by wentaoxandry** [espnet/espnet#1420]
Hello, 
I'm interested in ESPnet with a new dataset, I have done some experiments, but I got bad WER.
The dataset, which I used is LRS2 corpus(large vocabulary).  I have tried training the example scripts from WJS, AMI, commonvoice, librispeech, but all got bad WER (over 50%). I want to ask if someone has an idea about that? Or how can I train the new dataset with ESPnet? 
What I have done in my experiments is only replace the training data. The training stage and the network config file haven't be changed. 
Best regard,
Wentao

**Comment by Fhrozen** [comment]
First to all, Thank you for using ESPnet.
Second, could you please share the link of the dataset (just to confirm is this: http://www.robots.ox.ac.uk/~vgg/data/lip_reading/lrs2.html).

You may need to be careful with the alignment of the audio with the text. Also, remember that the each recipe has its own way to prepare the lists (tag_of_each_uttr) and that this may be wrong in the output. As example, you can compare the one in https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/local/data_prep.sh librispeech and the one in https://github.com/espnet/espnet/blob/master/egs/chime5/asr1/local/prepare_data.sh chime5 and are different.

Try to test with a small dataset, such as AN4 to understand how the data files are prepared. Also try to check your json file, located in the dump files.

If you could share the dataset link (if it is public), I could check if there is an additional prepare data to be coded.

**Comment by wentaoxandry** [comment]
Hello Fhrozen, 
thank you for your reply!
Yes, the link of the dataset is http://www.robots.ox.ac.uk/~vgg/data/lip_reading/lrs2.html
For the alignment of the audio with the text, I think there is nothing wrong because I also trained the dataset with Kaldi script, which can get good results. 
I also tried the AN4, and compared the json files with my files. It seems nothing wrong.
Thank you for your advice. If I want to train with a large vocabulary dataset, which example script will you recommend as the baseline?

**Comment by Fhrozen** [comment]
Thank you for your quick response.
The larger I worked with were librispeech and tedlium 3. 

Maybe you can try with librispeech  largest model.

**Comment by wentaoxandry** [comment]
Thank you, I will try it.
Now I'm trying with librispeech, I do not have 4 gpus so I reduce gpu number and reduce the batch-bins. Here are the training results, But I think something is wrong during training? Or I should run all epochs and see the decoding results?
Thank you
![loss](https://user-images.githubusercontent.com/47325204/70420204-78459080-1a67-11ea-92a7-ebcc1423efcc.png)
![acc](https://user-images.githubusercontent.com/47325204/70420205-78459080-1a67-11ea-9488-8e570238e81a.png)

**Comment by Fhrozen** [comment]
Mhh, it does not look good. 
Right now I cannot try it until get a copy of the LRS (need to get the agreement).

You see than the training is not good from the accuracy. Usually at epoch 8/10 it should be around 0.5-0.6 with large corpus. for small corpus (WSJ) it usually can reach 0.8 of val acc at epoch 10.

Will need to check the data once I get a copy.
In this case, try with the RNN to see if the training is also failing. Use verbose 1 to check the training acc at the output. Also activate the cer/wer calculation during training. use:

```
report-cer: true
report-wer: true
```
in your model config file.

**Comment by wentaoxandry** [comment]
Thank you for your advice, I will try it

**Comment by sw005320** [comment]
Thanks for your report!
The data look very interesting. It would be great if you consider making a PR for your recipe.
We'll support you.

Can you show me your config file?
We usually do not have such training/validation loss/acc degradation, and this is due to some wrong setup in the optimization-related hyperparameter.

**Comment by wentaoxandry** [comment]
@sw005320 
Thank you for your support, here is config file which I used :
# librispeech
accum-grad: 4
adim: 512
aheads: 8
backend: pytorch
batch-bins: 1500000
dlayers: 6
dropout-rate: 0.1
dunits: 2048
elayers: 12
epochs: 120
eunits: 2048
grad-clip: 5
lsm-weight: 0.1
model-module: espnet.nets.pytorch_backend.e2e_asr_transformer:E2E
mtlalpha: 0.3
opt: noam
patience: 10
sortagrad: 0
transformer-attn-dropout-rate: 0.0
transformer-init: pytorch
transformer-input-layer: conv2d
transformer-length-normalized-loss: false
transformer-lr: 10.0
transformer-warmup-steps: 25000

I also tried AMI script, which has best results until now:
# network architecture
# encoder related
elayers: 12
eunits: 2048
# decoder related
dlayers: 6
dunits: 2048
# attention related
adim: 256
aheads: 4

# hybrid CTC/attention
mtlalpha: 0.2

# label smoothing
lsm-weight: 0.1

# minibatch related
batch-size: 32
maxlen-in: 512  # if input length  > maxlen-in, batchsize is automatically reduced
maxlen-out: 150 # if output length > maxlen-out, batchsize is automatically reduced

# optimization related
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
opt: noam
accum-grad: 2
grad-clip: 5
patience: 10
epochs: 200
dropout-rate: 0.1

# transformer specific setting
backend: pytorch
model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"
transformer-input-layer: conv2d     # encoder architecture type
transformer-lr: 10.0
transformer-warmup-steps: 25000
transformer-attn-dropout-rate: 0.0
transformer-length-normalized-loss: false
transformer-init: pytorch 

with AMI script, I got the WER 37.7%. But compare with Kaldi result (10.1%), I think there is something wrong in my confile. 
By the way I use LRS2 both pretrain and train set for the model training.

Hearty greetings,
Wentao and Dorothea

**Comment by sw005320** [comment]
Can you decrease `transformer-lr`? (e.g., 5.0, 2.0, 1.0 etc.)?
Then, I guess you'd not have some degradations in the early epoch stage.
Also, you may increase `accum-grad` later.

> with AMI script, I got the WER 37.7%. But compare with Kaldi result (10.1%), I think there is something wrong in my confile.

The e2e model does not outperform the Kaldi especially in this range of the training data.
I think this is a reasonable number.

@ruizhilijhu, the data seem to be very interesting and you may also take a look at this.

**Comment by wentaoxandry** [comment]
@sw005320 , thank you for your advice, I will try it. If I got the results, I will tell you as soon as possible.
I'm have read a paper https://ibug.doc.ic.ac.uk/media/uploads/documents/av_speech_hybridctcattention.pdf. In this paper with LRS2 and ESPnet, they got WER 8.3 for audio-only. I have tried to connect the authors, but haven't got reply. I'm not sure which script or config file they used, maybe this paper is also interesting for you.

**Comment by Fhrozen** [comment]
There are two parts you need to consider first. (Not sure if for kaldi is also required)
Language model:
```
The language model is trained by combining two different
text corpora. The first one contains the transcriptions of the
LibriSpeech corpus which contains 9.4 million words. The
second one contains the transcriptions of the LRS2 pre-train
set which contains more than 2 million words.
```

Data Augmentation:
```
Data augmentation is applied to the raw audio sequences before computing
the mel and pitch features. During training babble noise at
different SNR levels (0 dB, 5 dB and 10 dB) from the NOISEX database [28] might be added to the original audio clip.
The selection of one of the noise levels or the use of the clean
audio is done using a uniform distribution.
```
Maybe the result without this two should be similar (close) to the one you have reported for kaldi (WER: 10)

**Comment by wentaoxandry** [comment]
@Fhrozen , thank you for your reply, I have also tried the language model without the LibriSpeech corpus, the WER does not change. For data augmentation, now I haven't do any data augmentation, so all audios are clean signals.

**Comment by wentaoxandry** [comment]
@sw005320 , if you want we can work together to make a baseline script for LRS2.

**Comment by sw005320** [comment]
Sure, let's do that. We can communicate through this issue or directly send an email to me (shinjiw@ieee.org) if you prefer it. Also, @ruizhilijhu (or others in JHU) would be interested in it.

**Comment by ruizhilijhu** [comment]
@wentaoxandry Hi, I am currently working on  multi-stream end-to-end ASR. Multimodal (visual+audio) ASR is one of the multi-stream scenarios. I am very interested in this dataset. If you'd like, we can work together to make a PR of this dataset on ESPnet. My email: ruizhili@jhu.edu

**Comment by tuanphan09** [comment]
![Screenshot from 2020-03-04 17-41-45](https://user-images.githubusercontent.com/30728127/75871973-f4216000-5e3f-11ea-8771-ebecebc389b7.png)

I'm training on a Vietnamese dataset, the ACC curve is pretty good but the validation loss (CTC loss)  seems meaningless. Is something wrong?

**Comment by sw005320** [comment]
I've never seen such trends. The training CTC loss looks OK, so the problem would be some issue in the validation transcriptions?

**Comment by kan-bayashi** [comment]
I faced similar CTC value when the length of transcription is shorter than that of inputs. It is better to check the length of transcription is longer than that of inputs / subsampling factor.

**Comment by joyolee** [comment]
Hello, I would ask if the LRS2 recipe is still under construction, or it will not be added?

**Comment by sw005320** [comment]
There are multiple requests and we'll make this a higher priority.
@LiChenda was working on it in https://github.com/espnet/espnet/commit/45318782d0a627374577648cec6ec82a60fc7dfb
@amadeusuzx also mentioned it.

@LiChenda, I know you're busy but maybe we can revisit it?
If we have some prototype PR, someone may take over.
Can we discuss it shortly at our meeting tomorrow?

**Comment by LiChenda** [comment]
Sure!

> There are multiple requests and we'll make this a higher priority.
> @LiChenda was working on it in [4531878](https://github.com/espnet/espnet/commit/45318782d0a627374577648cec6ec82a60fc7dfb)
> @amadeusuzx also mentioned it.
> 
> @LiChenda, I know you're busy but maybe we can revisit it?
> If we have some prototype PR, someone may take over.
> Can we discuss it shortly at our meeting tomorrow?

**Comment by wentaoxandry** [comment]
Hello Mr. Watanabe, 
now we are optimizing our audio-visual speech recognition egs script (which is based on LRS3 and LRS3 Corpus). Our paper 'Fusing information streams in end-to-end audio-visual speech recognition' (https://arxiv.org/abs/2104.09482) is published by ICASSP 2021.  We are nearly finished with the data processing codes (data segmentation, data augmentation) and next week we will optimize the PyTorch training codes for our audio-visual model. Would you like to work together on this task? I think it will be exciting for a collaboration
With sincerely regards 
Wentao

**Comment by sw005320** [comment]
Year, that is really cool.
Dorothea also mentioned your work.
@LiChenda also has some preliminary systems for LRS2, so we can have some coordination.
Maybe, we can have some meetings in the week after the next week (after the ASRU deadline)?
Please send me an email (shinjiw@ieee.org).

**Comment by LiChenda** [comment]
FYI, I have made a [PR]( https://github.com/espnet/espnet/pull/3346) for the LRS2 lipreading recipe.
